File: LeetCode/Liang/86PartitionList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *smaller = new ListNode(INT_MIN);
        ListNode *bigger = new ListNode(INT_MIN);
        ListNode *sTail = smaller, *bTail = bigger;
        while (head) {
            ListNode *tmp = head->next;
            if (head->val < x) {
                sTail->next = head;
                sTail = sTail->next;
            } else {
                bTail->next = head;
                bTail = bTail->next;
            }
            head = tmp;
        }
        sTail->next = bigger->next;
        head = smaller->next;
        bTail->next = nullptr;//Don't forget this.
        delete bigger;
        delete smaller;
        return head;
    }
};

File: LeetCode/Kai/86PartitionList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode dummy = new ListNode(0);
        ListNode newListDummy = new ListNode(0);
        dummy.next = head;
        ListNode curr = dummy;
        ListNode newCurr = newListDummy;
        while(curr.next!=null) {
            if(curr.next.val<x) curr = curr.next;
            else {
                newCurr.next = curr.next;
                curr.next = curr.next.next;
                newCurr.next.next = null;
                
                newCurr = newCurr.next;
            }
        }
        curr.next = newListDummy.next;
        return dummy.next;
        
    }
    
}